File: LeetCode/Study/ArrayAndString/11.ReverseString.py
"""
编写一个函数，其作用是将输入的字符串反转过来。输入字符串以字符数组 s 的形式给出。
不要给另外的数组分配额外的空间，你必须原地修改输入数组、使用 O(1) 的额外空间解决这一问题。
"""

from typing import List

class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        left, right = 0, len(s) - 1

        while left < right:
            s[left], s[right] = s[right], s[left]
            left+=1
            right-=1

def test():
    sol = Solution()
    test_cases = [
        (["h", "e", "l", "l", "o"], ["o", "l", "l", "e", "h"]),
        (["a", "b", "c", "d", "e"], ["e", "d", "c", "b", "a"]),
        (["r", "a", "c", "e", "c", "a", "r"], ["r", "a", "c", "e", "c", "a", "r"]),  # palindrome
        (["a", "b", "c"], ["c", "b", "a"]),
        (["x", "y", "z"], ["z", "y", "x"]),
        (["a"], ["a"]),  # single character (edge case)
        ([], []),  # empty string (edge case)
    ]
    
    for i, (input_str, expected) in enumerate(test_cases, 1):
        sol.reverseString(input_str)
        print(f"Test case {i}:")
        print(f"Input: {input_str}")
        print(f"Expected: {expected}, Result: {input_str}")
        print("✅ Pass" if input_str == expected else "❌ Fail")
        print("-" * 40)

if __name__ == "__main__":
    test()

File: LeetCode/Study/ArrayAndString/11.ReverseString.java
/*
 * 编写一个函数，其作用是将输入的字符串反转过来。输入字符串以字符数组 s 的形式给出。
 * 不要给另外的数组分配额外的空间，你必须原地修改输入数组、使用 O(1) 的额外空间解决这一问题。
 */

import java.util.Arrays;

class Solution {
    public void reverseString(char[] s) {
        int left = 0;
        int right = s.length - 1;
        
        while (left < right) {
            char temp = s[left];
            s[left] = s[right];
            s[right] = temp;
            left++;
            right--;
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        // 用二维数组表示多个测试用例，输入字符数组以及期望的输出
        Object[][] testCases = {
            {new char[]{'h', 'e', 'l', 'l', 'o'}, new char[]{'o', 'l', 'l', 'e', 'h'}},
            {new char[]{'H', 'a', 'n', 'n', 'a', 'h'}, new char[]{'h', 'a', 'n', 'n', 'a', 'H'}},
            {new char[]{'a', 'b', 'c'}, new char[]{'c', 'b', 'a'}},
            {new char[]{'a'}, new char[]{'a'}}, // 单一字符，反转后应该还是原来
            {new char[]{}, new char[]{}} // 空字符数组
        };

        for (int i = 0; i < testCases.length; i++) {
            char[] input = (char[]) testCases[i][0];
            char[] expected = (char[]) testCases[i][1];
            sol.reverseString(input);

            System.out.printf("Test %d: input=%s\n", i + 1, new String(input));
            System.out.printf("Expected: %s, Result: %s\n", new String(expected), new String(input));
            boolean passed = Arrays.equals(input, expected);
            System.out.println(passed ? "✅ Pass" : "❌ Fail");
            System.out.println("-".repeat(40));
        }
    }
}
